proposals/0177-program-runtime-abiv2.md

```diff
@@ -0,0 +1,266 @@
+    - `InstructionAccount`, consisting of:
+      - Index to transaction account: `u16`
+      - Flags bitfield: `u16` (bit 0 is signer, bit 1 is writable)
+
+### Accounts area
+
+For each unique (meaning deduplicated) instruction account the payload must
+be mapped in at `0x800000000` plus `0x100000000` times the index of the
+**transaction** account (not the index of the instruction account). Only if the
+instruction account has the writable flag set and is owned by the current
+program it is mapped in as a writable region. The writability of a region must
+be updated as programs through out the transaction modify the account metadata.
+
+The runtime must only map the payload for accounts that belong in the current
+executing instruction. The payload for accounts belonging to sibling instructions
+must NOT be mapped.
+
+### Instruction payload area
+
+For each instruction, the runtime must map its payload at address 
+`0x10800000000` plus `0x100000000` times the index of the instruction in the 
+trasaction. All instruction payload mappings are readonly.
+
+One extra writable mapping must be created after the last instruction payload 
+area to be the CPI scratch pad, i.e. at address `0x10800000000` plus 
+`0x100000000` times the number of instructions in the transaction. Its purpose 
+is for programs to write CPI instruction data directly to it and avoid copies.
+
+### VM initialization
+
+During the initilization of the virtual machine, the runtime must load the 
+value `0x400000000` in register `R1` and value `0x600000000` in register `R2`. 
+These values represent addresses for programs to easily find the areas to read 
+information about the transaction and the instructions.
+
+### Changes to syscalls
+
+Changes to the account metadata must now be communicated with specific 
+syscalls, as detailed below:
+
+- `sol_resize_account`: Dst account, new length as `u64`
+- `sol_assign_owner`: Dst account, new owner as `&[u8; 32]`
+- `sol_transfer_lamports`: Dst account, src account, amount as `u64`
+
+The account parameters are the index of the account in the transaction.
+
+Changes to the account payload length and all the scratchpads sections 
+introduced in this SIMD (the return-data scratchpad and the CPI scratchpad) 
+must be communicated via a new sycall `set_buffer_length`, with the following 
+parameters:
+
+- Address of region to be resized: `u64`
+- New length of region: `u64`
+
+The syscall must check if the address belongs to either a writable account 
+payload or one of the scratchpads and return and error otherwise. Constrains 
+for the maximum resizable limits must also be verified (10 kb).
+
+The verifier must reject SBPFv4 programs containing the `sol_invoke_signed_c` 
+and `sol_invoke_signed_rust`, since they are not compatible with ABIv2. A new 
+syscall `sol_invoke_signed_v2` must replace them. The parameters for 
+`sol_invoke_signed_v2` are the following:
+
+- Index in transaction of program ID to be called: `u64`.
```

Comment:
Could this be a reference to the program id, as it is now? Otherwise it seems that a program will need to lookup the index for the callee program.

Reply:
Programs will have access to an array containing all accounts in the transaction. They will be aware of the index of the accounts the instruction is referring to. Having said that, the index is easily accessible and easier to manage.